### PR TITLE
Feature/validation status

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/blob.hpp [hunk 1]
```diff
@@ -61,6 +57,19 @@ struct Blob {
     return *this;
   }
 
+  int compareTo(const Blob<N> b) const {
+    for (int i = N - 1; i >= 0; --i) {
```
Comment (Warchant) [inline include/veriblock/blob.hpp]: you can use `memcmp` instead

File: include/veriblock/blob.hpp [hunk 2]
```diff
@@ -61,6 +57,19 @@ struct Blob {
     return *this;
   }
 
+  int compareTo(const Blob<N> b) const {
```
Comment (Warchant) [inline include/veriblock/blob.hpp]:
```suggestion
  int compareTo(const Blob<N>& b) const {
```

File: include/veriblock/entities/btcblock.hpp [hunk 3]
```diff
@@ -47,6 +48,19 @@ struct BtcBlock {
...
+  Sha256Hash getBlockHash() const {
+    Sha256Hash hash;
+
+    WriteStream stream;
+    toRaw(stream);
+
+    sha256_context context;
+    sha256_init(&context);
+    sha256_update(&context, stream.data().data(), stream.data().size());
+    sha256_finish(&context, hash.data());
```
Comment (Warchant) [inline include/veriblock/entities/btcblock.hpp]: use 
```
#include "veriblock/hashutil.hpp"

...
return sha256(stream.data());
```

File: include/veriblock/entities/hashes.hpp [hunk 4]
```diff
@@ -11,6 +11,34 @@ using Sha256Hash = Blob<SHA256_HASH_SIZE>;
 using VBlakePrevKeystoneHash = Blob<VBLAKE_PREVIOUS_KEYSTONE_SIZE>;
 using VBlakeBlockHash = Blob<VBLAKE_PREVIOUS_BLOCK_SIZE>;
 
+inline Sha256Hash decodeBits(const uint32_t& bits) {
+  std::vector<uint8_t> target(SHA256_HASH_SIZE);
```
Comment (Warchant) [inline include/veriblock/entities/hashes.hpp]:
```suggestion
  Sha256Hash target;
```

Comment (Warchant) [reply include/veriblock/entities/hashes.hpp]: pls rename Sha256Hash to uint256

File: src/stateless_validation.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+bool checkProofOfWork(const BtcBlock& block) {
+  Sha256Hash blockHash = block.getBlockHash();
+  Sha256Hash target = decodeBits(block.bits);
+
+  return target > blockHash;
+}
+
+bool checkBtcBlock(const BtcBlock& block, ValidationState& state) {
+  if (!checkProofOfWork(block)) {
+    return state.Invalid("checkBtcBlock is failed",
```
Comment (Warchant) [inline src/stateless_validation.cpp]:
```suggestion
    return state.Invalid("checkBtcBlock",
```

File: include/veriblock/entities/btcblock.hpp [hunk 6]
```diff
@@ -55,10 +58,9 @@ struct BtcBlock {
     WriteStream stream;
     toRaw(stream);
 
-    sha256_context context;
-    sha256_init(&context);
-    sha256_update(&context, stream.data().data(), stream.data().size());
-    sha256_finish(&context, hash.data());
+    sha256(hash.data(), stream.data().data(), (uint32_t)stream.data().size());
+    sha256(hash.data(), hash.data(), SHA256_HASH_SIZE);
```
Comment (Warchant) [inline include/veriblock/entities/btcblock.hpp]: use a `sha256d` from `hashutil.hpp`

File: include/veriblock/entities/btcblock.hpp [hunk 7]
```diff
@@ -20,8 +21,10 @@ struct BtcBlock {
   static BtcBlock fromRaw(ReadStream& stream) {
     BtcBlock block{};
     block.version = stream.readLE<uint32_t>();
-    block.previousBlock = stream.readSlice(SHA256_HASH_SIZE);
-    block.merkleRoot = stream.readSlice(SHA256_HASH_SIZE);
+    block.previousBlock =
+        ((Sha256Hash)stream.readSlice(SHA256_HASH_SIZE)).reverse();
+    block.merkleRoot =
```
Comment (Warchant) [inline include/veriblock/entities/btcblock.hpp]: Semantically it is wrong to cast a slice to Sha256Hash.
I think better would be to add method `std::vector<uint8_t> reverse() const noexcept` to a Slice

File: include/veriblock/consts.hpp [hunk 8]
```diff
@@ -12,6 +14,7 @@ constexpr const auto MAX_HEADER_SIZE_PUBLICATION_DATA = 1024;
 constexpr const auto MAX_PAYOUT_SIZE_PUBLICATION_DATA = 100;
 constexpr const auto MAX_CONTEXT_SIZE_PUBLICATION_DATA = 100;
 constexpr const auto SHA256_HASH_SIZE = 32;
+constexpr const auto VBLAKE_BLOCK_SIZE = 24;
```
Comment (Warchant) [inline include/veriblock/consts.hpp]: it is bad name for a const, because actually VBK_BLOCK_SIZE is 64 bytes, but here you mean HASH of block. 
Better name it `VBLAKE_BLOCK_HASH_SIZE`

File: include/veriblock/consts.hpp [hunk 9]
```diff
@@ -1,6 +1,8 @@
 #ifndef ALT_INTEGRATION_INCLUDE_VERIBLOCK_CONSTS_HPP_
 #define ALT_INTEGRATION_INCLUDE_VERIBLOCK_CONSTS_HPP_
 
+#include <stdint.h>
```
Comment (Warchant) [inline include/veriblock/consts.hpp]:
```suggestion
#include <cstdint>
```

File: include/veriblock/blob.hpp [hunk 10]
```diff
@@ -31,13 +31,13 @@ struct Blob {
...
+  iterator begin() noexcept { return data_.data(); }
 
-  const_iterator begin() const noexcept { return data_.begin(); }
+  const_iterator begin() const noexcept { return data_.data(); }
 
-  iterator end() noexcept { return data_.end(); }
+  iterator end() noexcept { return data_.data() + N; }
 
-  const_iterator end() const noexcept { return data_.end(); }
+  const_iterator end() const noexcept { return data_.data() + N; }
```
Comment (Warchant) [inline include/veriblock/blob.hpp]: why did you change all this?

File: include/veriblock/entities/vbkblock.hpp [hunk 11]
```diff
@@ -5,32 +5,33 @@
...
-  int32_t difficulty{};
-  int32_t nonce{};
+  uint32_t timestamp{};
+  uint32_t difficulty{};
+  uint32_t nonce{};
 
   static VbkBlock fromRaw(ReadStream& stream) {
     VbkBlock block{};
-    block.height = stream.readBE<int32_t>();
+    block.height = stream.readBE<uint32_t>();
```
Comment (Warchant) [inline include/veriblock/entities/vbkblock.hpp]: but block height is signed in java impl?

File: include/veriblock/entities/vbkblock.hpp [hunk 12]
```diff
@@ -5,32 +5,33 @@
...
     block.previousKeystone = stream.readSlice(VBLAKE_PREVIOUS_KEYSTONE_SIZE);
     block.secondPreviousKeystone =
         stream.readSlice(VBLAKE_PREVIOUS_KEYSTONE_SIZE);
     block.merkleRoot = stream.readSlice(VBK_MERKLE_ROOT_SIZE);
-    block.timestamp = stream.readBE<int32_t>();
-    block.difficulty = stream.readBE<int32_t>();
-    block.nonce = stream.readBE<int32_t>();
+    block.timestamp = stream.readBE<uint32_t>();
+    block.difficulty = stream.readBE<uint32_t>();
+    block.nonce = stream.readBE<uint32_t>();
```
Comment (Warchant) [inline include/veriblock/entities/vbkblock.hpp]: timestamp, difficulty, nonce -- all are signed in VBK in java?

File: include/veriblock/entities/vbkblock.hpp [hunk 13]
```diff
@@ -58,6 +59,17 @@ struct VbkBlock {
     toRaw(blockStream);
     writeSingleByteLenValue(stream, blockStream.data());
   }
+
+  VBlakeBlockHash getBlockHash() const {
```
Comment (Warchant) [inline include/veriblock/entities/vbkblock.hpp]:
```suggestion
  VBlakeBlockHash getHash() const {
```

File: include/veriblock/entities/hashes.hpp [hunk 14]
```diff
@@ -9,9 +9,11 @@ namespace VeriBlock {
 using VbkMerkleRootSha256Hash = Blob<VBK_MERKLE_ROOT_SIZE>;
 using Sha256Hash = Blob<SHA256_HASH_SIZE>;
 using VBlakePrevKeystoneHash = Blob<VBLAKE_PREVIOUS_KEYSTONE_SIZE>;
-using VBlakeBlockHash = Blob<VBLAKE_PREVIOUS_BLOCK_SIZE>;
+using VBlakePrevBlockHash = Blob<VBLAKE_PREVIOUS_BLOCK_SIZE>;
+using VBlakeBlockHash = Blob<VBLAKE_BLOCK_SIZE>;
 
-inline Sha256Hash decodeBits(const uint32_t& bits) {
+template <size_t N>
```
Comment (Warchant) [inline include/veriblock/entities/hashes.hpp]: why template?

File: include/veriblock/entities/hashes.hpp [hunk 15]
```diff
@@ -9,9 +9,11 @@ namespace VeriBlock {
...
+inline Blob<N> decodeBits(const uint32_t& bits) {
```
Comment (Warchant) [inline include/veriblock/entities/hashes.hpp]: move this function to a veriblock/blockutil.hpp

File: src/stateless_validation.cpp [hunk 16]
```diff
@@ -1,23 +1,57 @@
...
+  VBlakeBlockHash blockHash = block.getBlockHash();
+  Blob<VBLAKE_BLOCK_SIZE> target =
+      decodeBits<VBLAKE_BLOCK_SIZE>(block.difficulty);
 
   return target > blockHash;
 }
 
+template <typename BlockType>
+bool checkMaximumDrift(const BlockType& block) {
+  uint32_t currentTime = (uint32_t)time(0);
```
Comment (Warchant) [inline src/stateless_validation.cpp]: create a file `veriblock/time.hpp` and create a function which returns current timestamp.
and use that func there.

better name it something like `currentTimestamp4` (stands for 4 bytes, because we could have another func which returns 8 bytes)

File: src/stateless_validation.cpp [hunk 17]
```diff
@@ -1,23 +1,57 @@
...
+}
+
 bool checkBtcBlock(const BtcBlock& block, ValidationState& state) {
   if (!checkProofOfWork(block)) {
     return state.Invalid("checkBtcBlock is failed",
                          "Block hash is higher than target");
   }
 
+  if (!checkMaximumDrift(block)) {
+    return state.Invalid("checkBtcBlock is failed",
```
Comment (Warchant) [inline src/stateless_validation.cpp]:
```suggestion
    return state.Invalid("checkBtcBlock failed",
```

File: include/veriblock/blob.hpp (hunk 10 above)
Comment (Mr-Leshiy) [reply include/veriblock/blob.hpp]: because data.begin() it is an ArrayIterator type and our 'iterator' is value_type*

File: include/veriblock/entities/vbkblock.hpp (hunk 12 above)
Comment (Mr-Leshiy) [reply include/veriblock/entities/vbkblock.hpp]: yes they are all signed, I will change this

File: include/veriblock/blockutils.hpp [hunk 18]
```diff
@@ -0,0 +1,32 @@
+#include "entities/hashes.hpp"
+
+namespace VeriBlock {
+
+inline uint256 decodeBits(const uint32_t& bits) {
+  std::vector<uint8_t> target(SHA256_HASH_SIZE);
```
Comment (Warchant) [inline include/veriblock/blockutils.hpp]:
```suggestion
  uint256 target;
```

File: test/sha256_test.cpp [hunk 19]
```diff
@@ -64,4 +66,80 @@ TEST_P(Sha256Test, IUF) {
...
+  ASSERT_TRUE(hash1 == hash2);
+}
+
+TEST(Sha256Test, uint256_decodeBits) {
+  uint256 target = decodeBits(0x04800000);
+
+  EXPECT_EQ(
+      target,
+      uint256(
+          "0000000000000000000000000000000000000000000000000000000000000000"_unhex));
```
Comment (Warchant) [inline test/sha256_test.cpp]: pls rewrite this test to be value-parametrized https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests

similar to vblake_test or sha256_test

```
struct TestCase {
  uint32_t bits;
  uint256 target;
}
```
